### ISSUE TITLE
Clarify that `compile(module)` only affects the `forward` method.

### ISSUE THREAD

**Issue by randolf-scholz** [pytorch/pytorch#141616]
### 📚 The doc issue

The following code silently does not trigger compilation, apparently because `forward` is not called. One can see this by using a debug flag or passing an illegal configuration like `backend="cudagraphs", mode="reduce-overhead"` to `compile`. 

Certain models like Normalizing Flows and distribution-like modules might be used without calling `forward`. 

```python
import os
os.environ["TORCH_LOGS"] = "dynamo"
import torch
from torch import Tensor, nn

class Flow(nn.Module):
    def forward(self, x: Tensor) -> Tensor:
        return x + 1.0

    def inverse(self, x: Tensor) -> Tensor:
        return x - 1.0

m = Flow().to(device="cuda")
optimized: Flow = torch.compile(m)    # type: ignore

x = torch.rand(1, 2, 3, 4).to(device="cuda")
y1 = m.inverse(x)
y2 = optimized.inverse(x)
assert torch.equal(y1, y2)
print("No compilation until here.")
optimized(x)  # <- triggers compilation of `forward`
```

### Suggest a potential alternative/fix

Clarify that when using `torch.compile(module)`, it is only triggered upon calling `forward`, and that only `forward` gets compiled. The current documentation is imprecise:

#### https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html

> In the same fashion, when compiling a module all sub-modules and methods within it, that are not in a skip list, are also compiled.

- Apparently, not "all" methods within the module get compiled, but only the ones `forward` depends on.
- Secondly, assume `MyModel.forward` calls `MyModel.helper_fn`. After compilation, if I call `helper_fn` in a standalone manner, do I get a compiled version? Or does it only compile it in an "inlined" manner?

> When you use `torch.compile`, the compiler will try to recursively compile every function call inside the target function or module inside the target function or module that is not in a skip list (such as built-ins, some functions in the torch.* namespace).

This should probably read *"inside the target module's forward method"* instead.

#### https://pytorch.org/docs/stable/generated/torch.compile.html

This should also mention that compilation only applies to forward.
